util/src/main/java/io/kubernetes/client/leaderelection/LeaderElector.java

```diff
@@ -0,0 +1,315 @@
+    logger.info("attempting to acquire leader lease {}...", desc);
+    try {
+      while (!cancel.get()) {
+        boolean success = tryAcquireOrRenew();
+        maybeReportTransition();
+        if (success) {
+          config.getLock().recordEvent("became leader");
+          logger.info("successfully acquired lease {}", desc);
+          return true;
+        } else {
+          logger.debug("failed to acuqire lease {}", desc);
+        }
+        // sleep time jittered
+        Thread.sleep(config.getRetryPeriod()
+            + (long) (random.nextDouble() * JITTER_FACTOR * config.getRetryPeriod()));
+      }
+    } catch (InterruptedException e) {
+      logger.warn("acquire thread interrupted", e);
+    }
+    return false;
+  }
+
+  /**
+   * renew loops calling tryAcquireOrRenew and returns immediately when tryAcquireOrRenew fails or
+   * ctx signals done.
+   */
+  private void renew(AtomicBoolean cancel) {
+    try {
+      while (!cancel.get()) {
+        String desc = config.getLock().describe();
+        try {
+          Callable<Boolean> callable = () -> {
+            boolean done = false;
+            while (!done && !cancel.get()) {
+              done = tryAcquireOrRenew();
+              Thread.sleep(config.getRetryPeriod());
+            }
+            return done;
+          };
+          executor.invokeAll(
+              Arrays.asList(callable), config.getRenewDeadline(), TimeUnit.MILLISECONDS);
+          maybeReportTransition();
+          logger.debug("successfully renewed lease {}", desc);
+        } catch (InterruptedException e) {
+          logger.warn("failed to tryAcquireOrRenew", e);
+        } catch (Exception e) {
+          config.getLock().recordEvent("stopped leading");
+          logger.info("failed to renew lease {}", desc, e);
+        }
+        Thread.sleep(config.getRetryPeriod());
+      }
+    } catch (InterruptedException e) {
+      logger.warn("renew thread interrupted", e);
+    } finally {
+      if (config.isReleaseOnCancel()) {
+        release();
+      }
+    }
+  }
+
+  /**
+   * tryAcquireOrRenew tries to acquire a leader lease if it is not already acquired, else it tries
+   * to renew the lease if it has already been acquired. Returns true on success else returns
+   * false.
+   *
+   * @return return true when acquire (or renew) lock success, otherwise false
+   */
+  private boolean tryAcquireOrRenew() {
+    final long now = System.currentTimeMillis();
+    LeaderElectionRecord leaderElectionRecord = new LeaderElectionRecord(
+        config.getLock().identity(),
+        (int) TimeUnit.MILLISECONDS.toSeconds(config.getLeaseDuration()), now, now);
+    LeaderElectionRecord oldLeaderElectionRecord;
+    try {
+      // 1. obtain or create the ElectionRecord
+      oldLeaderElectionRecord = config.getLock().get();
+    } catch (ResourceNotFoundException e) {
+      try {
+        config.getLock().create(leaderElectionRecord);
+      } catch (ApiException e1) {
```

Comment:
a better name for `e1`. we don't like to do extra reverse engineering work, do we?